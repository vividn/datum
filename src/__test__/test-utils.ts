import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
