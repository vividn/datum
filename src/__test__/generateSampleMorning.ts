import { endCmd } from "../commands/endCmd";
import { occurCmd } from "../commands/occurCmd";
import { startCmd } from "../commands/startCmd";
import { switchCmd } from "../commands/switchCmd";
import { popNow, pushNow, setNow } from "./test-utils";

export async function generateSampleMorning(date: string): Promise<void> {
  pushNow(`8:30 ${date}`);
  await endCmd("sleep");
  setNow("+5");
  await startCmd("sleep -y -t 2320");
  setNow("8:40");
  await switchCmd("project german");
  await switchCmd("text fiction_book");
  setNow("9:10");
  await endCmd("project");
  await endCmd("text");
  setNow("9:30");
  await switchCmd("environment outside");
  await startCmd("stretch");
  setNow("+7");
  await endCmd("stretch");
  await startCmd("run");
  setNow("+30");
  await endCmd("run distance=5.4");
  await startCmd("stretch");
  setNow("+8");
  await endCmd("stretch");
  await switchCmd("environment home");
  setNow("+3");
  await occurCmd("pushup amount=10");
  setNow("10:30");
  await switchCmd("environment outside 5");
  setNow("11");
  await occurCmd("caffeine amount=100 -c coffee");
  popNow();
}
