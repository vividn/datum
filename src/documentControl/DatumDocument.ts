import {
  isIsoDateOrTime,
  isoDate,
  isoDatetime,
  isoDuration,
} from "../time/timeUtils";
import { WithRequired } from "../utils/utilityTypes";

export type GenericData<T = unknown> = T & {
  [key: string]: any;
};

export type DatumData<T = unknown> = GenericData<T> & {
  occurTime?: isoDatetime | isoDate;
  occurUtcOffset?: number;
  dur?: "start" | "end" | isoDuration;
};

export type OccurredData<T = unknown> = WithRequired<
  DatumData<T>,
  "occurTime" | "occurUtcOffset"
>;

export function isOccurredData(
  data: DatumData | OccurredData
): data is OccurredData {
  const occurTime = data.occurTime;
  return occurTime !== undefined && isIsoDateOrTime(occurTime);
}

export type DatumMetadata = {
  utcOffset?: number;
  createTime?: isoDatetime;
  modifyTime?: isoDatetime; //TODO: turn into an array of times
  idStructure?: string;
  random?: number;
  humanId?: string;
  // [key: string]: any;
};

export type DatumPayload<T = unknown> = {
  _id?: string;
  _rev?: string;
  data: DatumData<T>;
  meta: DatumMetadata;
};

export type DatumDocument<T = unknown> = DatumPayload<T> & {
  _id: string;
  _rev: string;
};

export type DataOnlyPayload<T = unknown> = {
  _id?: string;
  _rev?: string;
} & DatumData<T>;

export type DataOnlyDocument<T = unknown> = DataOnlyPayload<T> & {
  _id: string;
  _rev: string;
};

export type EitherPayload<T = unknown> = DatumPayload<T> | DataOnlyPayload<T>;
export type EitherDocument<T = unknown> =
  | DatumDocument<T>
  | DataOnlyDocument<T>;

export function isDatumDocument(
  doc: DatumDocument | DataOnlyDocument
): doc is DatumDocument {
  return (
    (doc as DatumDocument).data !== undefined &&
    (doc as DatumDocument).meta !== undefined
  );
}

export function isDatumPayload(
  payload: DatumPayload | DataOnlyPayload
): payload is DatumPayload {
  return (
    (payload as DatumPayload).data !== undefined &&
    (payload as DatumPayload).meta !== undefined
  );
}
