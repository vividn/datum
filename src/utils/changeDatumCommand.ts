import { DatumData } from "../documentControl/DatumDocument";
import set from "lodash.set";
import { DataArgs } from "../input/dataArgs";
import { normalizeState } from "../state/normalizeState";
import get from "lodash.get";
import unset from "lodash.unset";
import { toDatumTime } from "../time/timeUtils";
import { DateTime } from "luxon";

// TODO: Use alterDatumData to change everything here rather than set/unset
export const commandChanges = ["occur", "start", "end", "switch"] as const;
export type CommandChange = (typeof commandChanges)[number];
// WARNING: Changes datumData and args in place
export function changeDatumCommand(
  datumData: DatumData,
  command: CommandChange,
  args?: DataArgs,
) {
  switch (command) {
    case "occur":
      datumData.occurTime ??= toDatumTime(DateTime.local());
      datumData.dur = null;
      break;

    case "start": {
      datumData.occurTime ??= toDatumTime(DateTime.local());
      const isActive = !!(
        datumData.state &&
        normalizeState(datumData.state) &&
        get(datumData, "state.id") !== false
      );
      if (!isActive) {
        set(datumData, "state.id", true);
      }
      if (datumData.dur === null) {
        unset(datumData, "dur");
      }
      if (Array.isArray(args?.optional)) {
        args!.optional.unshift("dur");
      }
      break;
    }

    case "end":
      datumData.occurTime ??= toDatumTime(DateTime.local());
      if (
        datumData.state === undefined ||
        normalizeState(datumData.state) !== false
      ) {
        if (
          datumData.state === undefined ||
          normalizeState(datumData.state) !== true
        ) {
          datumData.lastState ??= datumData.state;
        }
        datumData.state = { id: false };
      }
      if (datumData.dur === null) {
        unset(datumData, "dur");
      }
      if (Array.isArray(args?.optional)) {
        args!.optional.unshift("dur");
      }
      break;

    case "switch":
      datumData.occurTime ??= toDatumTime(DateTime.local());
      set(datumData, "state", {});
      if (datumData.dur === null) {
        unset(datumData, "dur");
      }
      if (Array.isArray(args?.optional)) {
        args!.optional.unshift("dur");
        args!.optional.unshift("state.id=true");
      }
      break;
  }
}
