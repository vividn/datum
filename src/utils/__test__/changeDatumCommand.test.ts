import { DatumData } from "../../documentControl/DatumDocument";
import { changeDatumCommand } from "../changeDatumCommand";
import { toDatumTime } from "../../time/timeUtils";
import { setNow } from "../../__test__/test-utils";
import { DataArgs } from "../../input/dataArgs";

const nowStr = "2023-12-20T17:00:00.000Z";
const occurTime = toDatumTime(nowStr);
setNow("2023-12-20T17:00:00.000Z");
describe("changeDatumCommand", () => {
  let datumData: DatumData;
  describe("start", () => {
    it("adds state.id true if state doesn't exist", () => {
      datumData = {
        existing: "data",
      };
      changeDatumCommand(datumData, "start");
      expect(datumData).toMatchObject({
        existing: "data",
        state: {
          id: true,
        },
      });
    });

    it("keeps a preexisting state", () => {
      datumData = {
        state: { id: "outside", weather: "rainy" },
        field: "environment",
      };
      changeDatumCommand(datumData, "start");
      expect(datumData).toMatchObject({
        state: { id: "outside", weather: "rainy" },
        field: "environment",
      });
    });

    it("keeps a prexisting state but changes state.id to true if state.id is false", () => {
      datumData = {
        field: "project",
        state: {
          id: false,
          countdown: 10,
        },
      };
      changeDatumCommand(datumData, "start");
      expect(datumData).toMatchObject({
        field: "project",
        state: {
          id: true,
          countdown: 10,
        },
      });
    });

    it("adds occurTime=now if it doesn't exist", () => {
      datumData = {
        existing: "data",
      };
      changeDatumCommand(datumData, "start");
      expect(datumData).toMatchObject({
        occurTime,
      });
    });

    it("keeps an existing occurTime", () => {
      const explicitOccur = toDatumTime("2023-12-21T11:45:00.000");
      datumData = {
        occurTime: explicitOccur,
      };
      changeDatumCommand(datumData, "start");
      expect(datumData).toMatchObject({
        occurTime: explicitOccur,
      });
    });

    it("removes dur=null from the data", () => {
      datumData = {
        state: { id: "something" },
        dur: null,
      };
      changeDatumCommand(datumData, "start");
      expect(datumData).toMatchObject({
        state: { id: "something" },
      });
      expect(datumData).not.toHaveProperty("dur");
    });

    it("adds a dur key to the beginning of the optional keys array", () => {
      datumData = {
        state: { id: "something" },
      };
      const args: DataArgs = { optional: ["existingKey"] };
      changeDatumCommand(datumData, "start", args);
      expect(args.optional).toEqual(["dur", "existingKey"]);
    });
  });

  describe("end", () => {
    it("adds state.id=false if state doesn't exist", () => {
      datumData = {
        existing: "data",
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toMatchObject({
        existing: "data",
        state: {
          id: false,
        },
      });
    });

    it("keeps state the same if it is false or { id: false }", () => {
      datumData = {
        state: { id: false },
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toMatchObject({
        state: { id: false },
      });
      expect(datumData).not.toHaveProperty("lastState");

      datumData = {
        state: false,
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toMatchObject({
        state: false,
      });
      expect(datumData).not.toHaveProperty("lastState");
    });

    it("turns a preexisting state into lastState and adds a state.id = false", () => {
      datumData = {
        state: { id: "outside", weather: "rainy" },
        field: "environment",
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toEqual({
        state: { id: false },
        field: "environment",
        lastState: { id: "outside", weather: "rainy" },
        occurTime,
      });
    });

    it("does not turn a simple state: true or state: {id: true} into lastState", () => {
      datumData = {
        state: { id: true },
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toEqual({
        state: { id: false },
        occurTime,
      });
      expect(datumData).not.toHaveProperty("lastState");

      datumData = {
        state: true,
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toEqual({
        state: { id: false },
        occurTime,
      });
      expect(datumData).not.toHaveProperty("lastState");
    });

    it("does not override existing lastState", () => {
      datumData = {
        state: { id: "inside" },
        lastState: { id: "outside", weather: "rainy" },
        field: "environment",
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toEqual({
        lastState: { id: "outside", weather: "rainy" },
        state: { id: false },
        field: "environment",
        occurTime,
      });
    });

    it("adds occurTime=now if it doesn't exist", () => {
      datumData = {
        existing: "data",
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toMatchObject({
        occurTime,
      });
    });

    it("keeps an existing occurTime", () => {
      const explicitOccur = toDatumTime("2023-12-21T11:45:00.000");
      datumData = {
        occurTime: explicitOccur,
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toMatchObject({
        occurTime: explicitOccur,
      });
    });

    it("removes dur=null from the data", () => {
      datumData = {
        state: { id: "something" },
        dur: null,
      };
      changeDatumCommand(datumData, "end");
      expect(datumData).toMatchObject({
        state: { id: false },
      });
      expect(datumData).not.toHaveProperty("dur");
    });
    it("adds a dur key to the beginning of the optional keys array", () => {
      datumData = {
        state: { id: "something" },
      };
      const args: DataArgs = { optional: ["existingKey"] };
      changeDatumCommand(datumData, "end", args);
      expect(args.optional).toEqual(["dur", "existingKey"]);
    });
  });

  describe("occur", () => {
    it("adds dur=null to the data", () => {
      datumData = {
        state: { id: "something" },
      };
      changeDatumCommand(datumData, "occur");
      expect(datumData).toMatchObject({
        state: { id: "something" },
        dur: null,
      });
    });
    it("overwrites an existing dur with null", () => {
      datumData = {
        state: { id: "something" },
        dur: "PT10M",
      };
      changeDatumCommand(datumData, "occur");
      expect(datumData).toMatchObject({
        state: { id: "something" },
        dur: null,
      });
    });

    it("adds occurTime=now if it doesn't exist", () => {
      datumData = {
        existing: "data",
      };
      changeDatumCommand(datumData, "occur");
      expect(datumData).toMatchObject({
        occurTime,
      });
    });
    it("keeps an existing occurTime", () => {
      const explicitOccur = toDatumTime("2023-12-21T11:45:00.000");
      datumData = {
        occurTime: explicitOccur,
      };
      changeDatumCommand(datumData, "occur");
      expect(datumData).toMatchObject({
        occurTime: explicitOccur,
      });
    });
  });

  describe("switch", () => {
    it("sets state to be {}", () => {
      datumData = {
        field: "field",
      };
      changeDatumCommand(datumData, "switch");
      expect(datumData).toMatchObject({
        field: "field",
        state: {},
      });
    });

    it("overwrites existing state with {}", () => {
      datumData = {
        state: { id: "something" },
        occurTime,
      };
      changeDatumCommand(datumData, "switch");
      expect(datumData).toEqual({
        state: {},
        occurTime,
      });
    });

    it("adds occurTime=now if it doesn't exist", () => {
      datumData = {
        existing: "data",
      };
      changeDatumCommand(datumData, "switch");
      expect(datumData).toMatchObject({
        occurTime,
      });
    });
    it("keeps an existing occurTime", () => {
      const explicitOccur = toDatumTime("2023-12-21T11:45:00.000");
      datumData = {
        occurTime: explicitOccur,
      };
      changeDatumCommand(datumData, "switch");
      expect(datumData).toMatchObject({
        occurTime: explicitOccur,
      });
    });

    it("removes dur=null from the data", () => {
      datumData = {
        state: { id: "something" },
        dur: null,
      };
      changeDatumCommand(datumData, "switch");
      expect(datumData).toEqual({
        state: {},
        occurTime,
      });
      expect(datumData).not.toHaveProperty("dur");
    });

    it("keeps an existing dur", () => {
      const explicitDur = "PT10M";
      datumData = {
        state: { id: "something" },
        dur: explicitDur,
        occurTime,
      };
      changeDatumCommand(datumData, "switch");
      expect(datumData).toEqual({
        state: {},
        dur: explicitDur,
        occurTime,
      });
    });

    it("adds a state=true and a dur key to the beginning of the optional keys array", () => {
      datumData = {
        state: { id: "something" },
      };
      const args: DataArgs = { optional: ["existingKey"] };
      changeDatumCommand(datumData, "switch", args);
      expect(args.optional).toEqual(["state.id=true", "dur", "existingKey"]);
    });
  });
});
