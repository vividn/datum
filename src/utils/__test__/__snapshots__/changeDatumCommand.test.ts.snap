// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changing from one command to another addCmd can become a start command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": true,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another addCmd can become a switch command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT5M30S",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": "stateName",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another addCmd can become an end command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": false,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another addCmd can become an occur command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": null,
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another endCmd can become a switch command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT5M30S",
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": "stateName",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another endCmd can become an occur command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": null,
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": false,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another endCmd can become start command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": true,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another occurCmd can become a start command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": true,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another occurCmd can become a switch command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT5M30S",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": "stateName",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another occurCmd can become an end command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": false,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another occurCmd doesn't require a duration to become a start command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": true,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another startCmd can become a switch command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT5M30S",
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": "stateName",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another startCmd can become an end command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": false,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another startCmd can become an occur command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": null,
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": true,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another switchCmd can become a start command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT5M30S",
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": "someState",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another switchCmd can become an end command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "key": "val",
    "lastState": "someState",
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": false,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`changing from one command to another switchCmd can become an occur command 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": null,
    "field": "field",
    "key": "val",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "state": "someState",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;
